azkaban-executor/azkaban-emr-jobtype/src/main/java/uk/gov/dwp/dataworks/azkaban/services/DependencyService.java

```diff
@@ -0,0 +1,161 @@
+package uk.gov.dwp.dataworks.azkaban.services;
+
+import com.amazonaws.services.dynamodbv2.AmazonDynamoDB;
+import com.amazonaws.services.dynamodbv2.model.AttributeValue;
+import com.amazonaws.services.dynamodbv2.model.GetItemRequest;
+import com.amazonaws.services.dynamodbv2.model.GetItemResult;
+import com.amazonaws.services.dynamodbv2.model.ScanRequest;
+import com.amazonaws.services.dynamodbv2.model.ScanResult;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.stream.Collectors;
+
+public class DependencyService extends DelegateService implements MetadataService {
+
+    public DependencyService(final AmazonDynamoDB dynamoDB, String metadataTableName, String exportDate) {
+        this.dynamoDb = dynamoDB;
+        this.metadataTableName = metadataTableName;
+        this.exportDate = exportDate;
+    }
+
+    public Optional<Map<String, AttributeValue>> successfulDependency(final String product) {
+        return proceed.get() ? dependencyMetadata(product).filter(this::dependencySucceeded)
+                                                          .flatMap(x -> dependencyMetadata(product)) : Optional.empty();
+    }
+
+    @Override
+    public void cancel() {
+        super.cancel();
+        this.latch.countDown();
+    }
+
+    private static Map<String, AttributeValue> primaryKey(final Map<String, AttributeValue> item) {
+        return item.entrySet().stream().filter(entry -> entry.getKey().equals(CORRELATION_ID_FIELD) || entry.getKey()
+                                                                                                            .equals(DATA_PRODUCT_FIELD))
+                   .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+    }
+
+    private Boolean dependencySucceeded(final Map<String, AttributeValue> item) {
+        return completedSuccessfully(metadataTableName, item).orElse(false);
+    }
+
+    private Optional<Boolean> completedSuccessfully(final String tableName, final Map<String, AttributeValue> item) {
+        try {
+            final AtomicBoolean succeeded = new AtomicBoolean(false);
+            final ScheduledExecutorService dependencyCheckExecutor = Executors.newSingleThreadScheduledExecutor();
```

Comment:
Is starting this is a separate thread and waiting for it to finish necessary? Why not just a simple loop?

Reply:
A simple loop could go on forever, this needs to timeout at some point. Adding a timeout to a simple loop arguably makes it less simple than this.